design/cluster-scope-resource-filter.md

```diff
@@ -0,0 +1,502 @@
+			- [all namespaced resources + some additional cluster resources](#all-namespaced-resources--some-additional-cluster-resources)
+			- [all namespaced resources + all cluster resources](#all-namespaced-resources--all-cluster-resources)
+			- [describe command change](#describe-command-change)
+	- [Detailed Design](#detailed-design)
+	- [Alternatives Considered](#alternatives-considered)
+	- [Security Considerations](#security-considerations)
+	- [Compatibility](#compatibility)
+		- [IncludedResources and ExcludedResources change](#includedresources-and-excludedresources-change)
+		- [IncludeClusterResources change](#includeclusterresources-change)
+	- [Implementation](#implementation)
+	- [Open Issues](#open-issues)
+
+## Abstract
+The current filter (IncludedResources/ExcludedResources + IncludeClusterResources flag) is not enough for some special cases, e.g. all namespaced resources + some kind of cluster resource and all namespaced resources + cluster resource excludes.
+Propose to separate `IncludedResources` into cluster scope and namespace scope parameters, and same change for `ExcludedResources`.
+
+## Background
+There are two sets of resource filters for Velero: `IncludedNamespaces/ExcludedNamespaces` and `IncludedResources/ExcludedResources`. 
+`IncludedResources` means only including the resource types specified in the parameter.
+The k8s resources are separated into cluster scope and namespaced scope.
+As a result, it's hard to include all resources in one group and only including specified resource in the other group.
+
+## Goals
+- Make the existing resource filters supporting more scenarios without conflict.
+
+## Non Goals
+- Enrich the resource filtering rules.
+
+
+## High-Level Design
+Four new parameters are added into command `velero backup create`: `--include-cluster-scope-resources`, `--exclude-cluster-scope-resources`, `--include-namespaced-resources` and `--exclude-namespaced-resources`. 
+`--include-cluster-scope-resources` and `--exclude-cluster-scope-resources` are used to filter cluster scope resources included or excluded in backup per resource type.
+`--include-namespaced-resources` and `--exclude-namespaced-resources` are used to filter namespace scope resources included or excluded in backup per resource type.
+
+### Parameters Rules
+
+* `--include-cluster-scope-resources` and `--include-namespaced-resources` valid value include `*` and comma separated string.
+
+* `--exclude-cluster-scope-resources` and `--exclude-namespaced-resources` valid value are comma separated string.
+
+* `--include-cluster-scope-resources`, `--include-namespaced-resources`, `--exclude-cluster-scope-resources` and `--exclude-namespaced-resources` parameters are mutual exclusive with `--include-resources` and `--exclude-resources` parameters. If both sets of parameters are provisioned, validation failure should be returned.
```

Comment:
@reasonerjt Yes. They serve different purposes. the --include/exclude-cluster-scope-resources overlaps with include/exclude-resources -- basically include/exclude is split into two categories. include-cluster-resources is a boolean pointer with 3 possible values which control how the include/exclude resources are processed.
1) nil: the default value. This means we only pull in cluster-scoped resources as needed by other resources (i.e. PVs, CRDs, etc.). This behavior is *modified* by includes/excludes -- so you could add CRD to exclude-cluster-scope-resources (or exclude-resources) which will result in related PVs being included but not CRDs
2) true: this means include cluster resources in general, but again modified by the includes/excludes. If includes/excludes include everything, then this means all CRDs are included. If you want to include all CRDs but not all ClusterRoleBindings, then you'd set `--include-cluster-resources --include-cluster-scope-resources=customresourcedefinitions` to include only CRDs or  `--include-cluster-resources --exclude-cluster-scope-resources=clusterrolebindings` to include all cluster resources *except* cluster role bindings.
3) false: this excludes all cluster resources

Reply:
@sseago per our discussion, for cluster-scope-resources we will always include those that are needed unless the user explicitly tells velero to exclude them from the backup, with this approach we may no longer need the 
bool flag `--include-cluster-resources`.  With that said, we should double-check the code to make sure this is doable with reasonable effort.

cc @blackpiglet

Reply:
@sseago 
Yes. Now, the backup has two resource filter parameters in the `backup.Request` structure: `NamespaceIncludesExcludes` and `ResourceIncludesExcludes`.

If we desperate `--include-cluster-resources` parameter, the `NamespaceIncludesExcludes` should have no change: if the namespace is listed in exclude array, it should be ignored, if there is a match in the include array, or include array contains an asterisk, the namespace is a keeper.
`ResourceIncludesExcludes` should be split into two parameters: 
* `NamespaceResourceIncludesExcludes`
* `ClusterResourceIncludesExcludes`. 

In the `backup` process, 
* The `NamespaceResourceIncludesExcludes` parameter should work the same way with `NamespaceIncludesExcludes`. 
* The `ClusterResourceIncludesExcludes` parameter should be kept by default, as long as it's not explicitly listed in the exclude array.

In the `restore` process,
* The `NamespaceResourceIncludesExcludes` should work the opposite as the backup process. Resources should be included by default unless it's in the excluded array.
* The `ClusterResourceIncludesExcludes` parameter works the same way as the backup.

Reply:
Not sure about the `Restore.Spec.RestoreStatus`. Looks like it needs no change.

Reply:
@blackpiglet 
"If we depreracte the --include-cluster-resources" 
From the last discussion, we shouldn't be deprecating this -- we should group this with the other existing params --include-resources and --exclude-resources. A user should either use some combination of the 3 old params and none of the new ones, or some combination of the four new params and none of the old ones. In other words, --include-cluster-resources is valid as long as the user doesn't also include any of the namespaced/cluster-scope params.

"The NamespaceResourceIncludesExcludes parameter should work the same way with NamespaceIncludesExcludes."

Almost, but not quite. NamespaceIncludesExcludes includes everything unless there's a specific resource list in includes, and then only those resources are included. So blank is the same thing as "*". For NamespaceResourceIncludesExcludes, blank means "don't include anything", while "*" means "include everything" -- this is to support the "don't include anything namespaced" use case.

"The ClusterResourceIncludesExcludes parameter should be kept by default, as long as it's not explicitly listed in the exclude array."
I don't think this is quite right. For cluster resources, there are 3 categories. Resources which are explicitly included will all be in the backup (i.e. if you include PVs, then *all* PVs are in the backup). Resources which are explicitly excluded will not be in the backup at all (if you exclude PVs, then *no* PVs are in the backup). Resources which are neither included nor excluded (missing in both lists) will only be pulled in if they are necessary for other items included. If PVs are in neither list, then we pull in PVs only if the PVC they're bound to is in the backup.

Reply:
> "If we depreracte the --include-cluster-resources"
> From the last discussion, we shouldn't be deprecating this -- we should group this with the other existing params --include-resources and --exclude-resources. A user should either use some combination of the 3 old params and none of the new ones, or some combination of the four new params and none of the old ones. In other words, --include-cluster-resources is valid as long as the user doesn't also include any of the namespaced/cluster-scope params.

IMO, we should consider deprecating the existing 3 params `--include-resources`/ `--exclude-resources` and  `--include-cluster-resources` from user interface if the new ones work well, we shouldn't be maintaining 2 sets of filters doing the same thing.

I find discussing the logic in English is more difficult than showing the code, therefore, if we reach an agreement regarding the behavior of the attributes, we should go ahead and merge this design and leave further implementation discussion in the code review.

In terms of delivering the features, if it's a bit challenge to deliver the new filters (--include/exclude-namespaced-resources, --include/exclude-cluster-scope-resources) for both backup and restore,  shall we try to deliver the new filters in backup first?  Because most problems are reported around backup.

@sseago @blackpiglet  what do you think?

Reply:
@reasonerjt I'm open to the idea, but I'd like to get input from users and customers I think. I'll reach out on the Red Hat side. As long as we go through a proper deprecation process (still in discussion, but sticking around for at least 2 releases after deprecation, etc.), there's no hard obstacle to removing them, and it would definitely simplify maintenance. It's possible that some users won't like having to think through separating namespaced/cluster-scoped resources, but thinking about it, as long as we get appropriate input before deprecating, this may be the best thing. It might be best to release the new params in a release first without deprecation, so people can experiment with both methods in practice and provide feedback before we make any final calls on deprecation.

Also, yes, I think it's reasonable to do this on backup only initially. Backup filtering is the most complicated, since we're working with everything in the cluster. On the restore side, it's just a matter of filtering on types that are already included in the backup. We may actually not want the new complexity on restore. Lets target backup side for initial implementation.

Reply:
@sseago 
Thanks for the comments. I will modify the document ASAP.
The agreement by far is keeping existing filters and creating new filters, and the new filters only work for backup resource handling.

Reply:
@blackpiglet That sounds good. If we do eliminate the existing filters, it will be in a future release after following the deprecation process, so for 1.11 the new and old filters must both work (with validation to make sure only one group is used for a given backup).